### ISSUE TITLE
easy way to generate fixed palette for taxa

### ISSUE THREAD

**Issue by david-barnett** [david-barnett/microViz#16]
- useful for comp_barplot (and ord_plot_iris)
- currently too tricky for a user to manually specify a palette, but being able to generate a template vector for further modification should help a lot
- should be easy enough with tax_top and distinct_palette
- could be called tax_palette?

**Comment by ChrisTrivedi** [comment]
Hi @david-barnett , 

First off, I love the microViz package so thanks a ton for sharing with the community!

I'm posting on this closed issue because I think it might be exactly the problem I am having. Please let me know if you'd prefer this to be its own issue and I'd be happy to create one. 

I have two datasets that I'm processing the same way but in different R projects. The end goal is to compare the bar plots between the two datasets. The issue is that taxonomy is called separately and they have slight differences when it comes to the most abundant taxa, so when using `comp_barplot` with the same n value I get different colors for the same taxa across the datasets. 

![image](https://user-images.githubusercontent.com/13127554/154708913-d6956cae-3551-42f4-82fa-3d77ec8c8424.png) ![image](https://user-images.githubusercontent.com/13127554/154709146-b29aa400-9dbd-4710-9a6b-c124e2e7b0b3.png)

The last thing I want to do is have to modify the colors manually in Inkscape, so I've tried a number of different ways around this like creating a custom palette that I can use with both, but the results are ugly...to say the least (because the palette has to be large enough to capture all the ASVs)! I've accomplished this so far purely by replacing my newly made palette in place of `distinct_palette(n=15, "brewerPlus")` in the `palette` flag.

![image](https://user-images.githubusercontent.com/13127554/154712847-f3b32152-826b-44d0-817f-6495e5708fdc.png) 
![image](https://user-images.githubusercontent.com/13127554/154713183-c994e253-6545-405f-9491-edcf20895c8c.png)

If there is an easier way to do this to keep the aesthetic of your `brewerPlus` palette that would be ideal. Any help would be greatly appreciated. Thanks in advance!

**Comment by david-barnett** [comment]
Hi Chris, glad you are finding microViz useful! 😃 

For any straightforward solution I could think of, you'll first need to put all your samples into one phyloseq (if not already done).

In some cases, you could use the group_by argument, which will make a list of plots, keeping the ordering and colours the same. With this you can also still use facets to group bars within each plot (seems like your plots have facets currently).

```r
dietswap %>%
  ps_filter(timepoint == 1, sex == "male") %>%
  comp_barplot(
    tax_level = "Family", n_taxa = 15,
    bar_outline_colour = NA, bar_width = 0.7, label = NULL,
    group_by = "nationality", facet_by = "bmi_group"
  )
```
![image](https://user-images.githubusercontent.com/33695702/154732406-c2564209-1895-464c-92f7-a5f7ec86c006.png)
![image](https://user-images.githubusercontent.com/33695702/154732444-241a5648-c732-4d22-a9f4-dcc5c08891ea.png)

But if you want to separate the ordering of the taxa from the palette assignment itself and/or have different legend levels per community, I thought that would be a little trickier, but it turns out it's not too bad.

I've defined a couple of functions below that I'll probably add to microViz soonish.

```r
# this makes a named palette vector from your combined dataset (considering overall abundance to assign colours)
tax_palette <- function(data, # phyloseq or ps_extra 
                        rank, # e.g. "Genus"
                        n, # n colours / taxa not including other
                        by = sum, # method for tax_sort
                        pal = "brewerPlus", # palette name from distinct_palette
                        add = c(other = "lightgrey"), # name = value pairs appended to end of output
                        ... # other args passed to tax_sort
                        ) {
  taxa <- tax_top(data = data, rank = rank, n = n, by = by, ...)
  taxColours <- distinct_palette(n = n, pal = pal, add = NA)
  
  names(taxColours) <- taxa
  taxColours <- c(taxColours, add)
  return(taxColours)
}

# palette viewer function (unnecessary but maybe handy)
tax_palette_plot <- function(
  named_pal_vec # named vector of colours
) {
  stopifnot(!anyNA(names(named_pal_vec))) # all colours need names
  p <- 
    named_pal_vec %>% 
    as.data.frame() %>% 
    tibble::rownames_to_column("taxon") %>% 
    dplyr::mutate(taxon = factor(taxon, levels = rev(taxon))) %>% 
    dplyr::rename("hex" = ".") %>% 
    ggplot2::ggplot(ggplot2::aes(y = taxon, fill = hex, x = "")) +
    ggplot2::geom_raster() +
    ggplot2::scale_fill_identity() + 
    ggplot2::labs(x = NULL, y = NULL) + 
    ggplot2::theme_minimal()
  return(p)
}
```

```r
myPal <- tax_palette(dietswap, rank = "Genus", pal = "brewerPlus", n = 40)
myPal %>% tax_palette_plot() # just to check the palette
```
![image](https://user-images.githubusercontent.com/33695702/154732911-312420cd-3d83-4406-864c-3b3ce7bd0de5.png)


```r
dietswap %>%
  ps_filter(nationality == "AFR", timepoint == 1, sex == "male") %>% # small subset of data
  comp_barplot(
    tax_level = "Genus", n_taxa = 15,
    bar_outline_colour = NA, bar_width = 0.7,
    palette = myPal, label = NULL
  ) 
```
![image](https://user-images.githubusercontent.com/33695702/154733182-2f6e858f-3f42-4a74-8d5e-0ab8c81cc1b6.png)

```r
dietswap %>%
  ps_filter(nationality != "AFR", timepoint == 1, sex == "male") %>% # different subset of data
  comp_barplot(
    tax_level = "Genus", n_taxa = 15,
    bar_outline_colour = NA, bar_width = 0.7,
    palette = myPal, label = NULL
  ) 
```
![image](https://user-images.githubusercontent.com/33695702/154733220-d2e60f66-0566-48cc-a3e7-f2a30144c690.png)

**Comment by ChrisTrivedi** [comment]
@david-barnett This is great, thanks so much for the info. My apologies for the delay in response. I'm trying your workaround now and will report back how it works. Thanks again!

**Comment by david-barnett** [comment]
included since 0.9.1 
https://david-barnett.github.io/microViz/news/index.html#microviz-091

**Comment by ChrisTrivedi** [comment]
> included since 0.9.1 https://david-barnett.github.io/microViz/news/index.html#microviz-091

@david-barnett my apologies for never getting back to you on how this worked for me. It was a resounding success and I want to thank you again. Also, cheers for adding it in the new update. To show my appreciation I also gave you a nod in the acknowledgments in our recent pub that used microViz - I hope that's okay with you. Thanks again!

**Comment by david-barnett** [comment]
thanks Chris! glad it worked for you and I appreciate the acknowledgement :)
